File: ui/src/lib/readingNotes.ts
import { invoke } from "@tauri-apps/api/core";
import DOMPurify from "dompurify";
import { assetBaseDirFor, renderBlock } from "./markdown";
import { validReadingNoteLabel } from "./readingNoteFormat";
import type { ReadingSelection, ReadingSelectionCapture } from "./readingSelection";

export interface ReadingNote {
  id: string;
  notePageId: string;
  filePath: string;
  body: string;
  revision: string;
  source: { pageId: string | null; pageTitle: string; filePath: string };
  quote: string;
  status: "attached" | "recovered" | "ambiguous" | "orphaned";
  statusMessage: string;
  targetBlockId: string | null;
  createdAt: string;
  updatedAt: string;
  storage: "inline" | "file";
  footnoteLabel: string | null;
  noteBlockId: string | null;
}

export interface ReadingNotesList {
  notes: ReadingNote[];
  warnings: string[];
}

export function readingNotesList(graphPath: string, pageId?: string): Promise<ReadingNotesList> {
  return invoke("reading_notes_list", { graphPath, ...(pageId ? { pageId } : {}) });
}

export function readingNoteCreate(
  graphPath: string, noteId: string, sourcePageId: string, selection: ReadingSelection | null, body: string,
): Promise<ReadingNote> {
  return invoke("reading_note_create", { graphPath, noteId, sourcePageId, selection, body });
}

export function readingNoteUpdate(
  graphPath: string, noteId: string, expectedRevision: string, body: string,
): Promise<ReadingNote> {
  return invoke("reading_note_update", { graphPath, noteId, expectedRevision, body });
}

export function readingNoteReattach(
  graphPath: string, noteId: string, expectedRevision: string, sourcePageId: string, selection: ReadingSelection,
): Promise<ReadingNote> {
  return invoke("reading_note_reattach", { graphPath, noteId, expectedRevision, sourcePageId, selection });
}

export function cloneReadingSelection(selection: ReadingSelection | null): ReadingSelection | null {
  if (!selection) return null;
  return {
    ...selection,
    blockIds: [...selection.blockIds],
    parts: selection.parts.map((part) => ({ ...part })),
    ...(selection.documentRange ? { documentRange: { ...selection.documentRange } } : {}),
  };
}

export function readingNoteSelection(captured: ReadingSelectionCapture, pageId: string): ReadingSelection | null {
  if (captured.pageIds.includes(pageId) && captured.error) throw new Error(captured.error);
  const selection = captured.selection;
  if (!selection || selection.pageId !== pageId) return null;
  if (!selection.text.trim() || !selection.parts.length || !selection.blockIds.length) {
    throw new Error("Select text in saved blocks on this page first.");
  }
  return cloneReadingSelection(selection);
}

/** DTO paths may be absolute; Markdown's renderer expects a graph-relative path. */
export function readingNoteRelativePath(graphPath: string, filePath: string): string {
  const graph = graphPath.replace(/\\/g, "/").replace(/\/+$/, "");
  const file = filePath.replace(/\\/g, "/");
  return file.startsWith(`${graph}/`) ? file.slice(graph.length + 1) : file;
}

export function readingNotePassageTarget(note: ReadingNote) {
  if (!["attached", "recovered"].includes(note.status) || !note.source.pageId || !note.targetBlockId) return null;
  return { pageId: note.source.pageId, pageName: note.source.pageTitle, targetBlockId: note.targetBlockId };
}

export function readingNoteLocationTarget(note: ReadingNote) {
  return note.noteBlockId
    ? { pageId: note.notePageId, pageName: note.source.pageTitle, targetBlockId: note.noteBlockId }
    : null;
}

export function renderReadingNoteBody(note: ReadingNote, graphPath: string): string {
  return renderReadingNoteMarkdown(note.body, assetBaseDirFor(readingNoteRelativePath(graphPath, note.filePath)));
}

export function renderReadingNoteMarkdown(body: string, baseDir: string): string {
  const html = renderBlock(body, baseDir);
  return DOMPurify.sanitize(html, {
    ALLOWED_URI_REGEXP: /^(?:https?:|mailto:|grafium-asset:|#|\/)/i,
    FORBID_TAGS: ["script", "style", "iframe", "object", "embed", "form", "input", "button"],
    // This is a read-only preview, not a page-link or task editor.
    FORBID_ATTR: ["style", "href", "data-page", "data-page-link", "data-tag", "data-ref", "data-block-ref", "data-reading-note-label", "autoplay"],
  });
}

export function renderReadingNoteFooter(body: string, label: string, baseDir: string): string {
  if (!validReadingNoteLabel(label)) return renderReadingNoteMarkdown(body, baseDir);
  const number = label.slice("grafium-note-".length);
  return `<section class="reading-note-footer" data-reading-note-footer="" aria-label="Reading note ${number}"><a class="reading-note-footer-label" href="#${label}" data-reading-note-label="${label}" aria-label="Open reading note ${number}">Note ${number}</a><div class="reading-note-footer-body">${renderReadingNoteMarkdown(body, baseDir)}</div></section>`;
}

File: ui/src-tauri/src/commands/reading_notes.rs
use grafium_core::graph::reading_notes::{ReadingNote, ReadingNotesList, ReadingSelection};
use grafium_core::Graph;
use std::path::Path;
use tauri::State;

fn current_graph(graph: &Graph, graph_path: &str) -> Result<(), String> {
    let mismatch = || "The active graph changed; reading notes were not accessed".to_string();
    if graph_path.trim().is_empty()
        || graph.root_dir.canonicalize().map_err(|_| mismatch())?
            != Path::new(graph_path)
                .canonicalize()
                .map_err(|_| mismatch())?
    {
        return Err(mismatch());
    }
    Ok(())
}

#[tauri::command(rename_all = "camelCase")]
pub fn reading_notes_list(
    state: State<crate::AppState>,
    graph_path: String,
    page_id: Option<String>,
) -> Result<ReadingNotesList, String> {
    let graph = state.graph.lock().map_err(|e| e.to_string())?;
    current_graph(&graph, &graph_path)?;
    graph
        .reading_notes_list(page_id.as_deref())
        .map_err(|e| e.to_string())
}

#[tauri::command(rename_all = "camelCase")]
pub fn reading_note_create(
    state: State<crate::AppState>,
    graph_path: String,
    note_id: String,
    source_page_id: String,
    selection: Option<ReadingSelection>,
    body: String,
) -> Result<ReadingNote, String> {
    let graph = state.graph.lock().map_err(|e| e.to_string())?;
    current_graph(&graph, &graph_path)?;
    graph
        .reading_note_create(&note_id, &source_page_id, selection.as_ref(), &body)
        .map_err(|e| e.to_string())
}

#[tauri::command(rename_all = "camelCase")]
pub fn reading_note_update(
    state: State<crate::AppState>,
    graph_path: String,
    note_id: String,
    expected_revision: String,
    body: String,
) -> Result<ReadingNote, String> {
    let graph = state.graph.lock().map_err(|e| e.to_string())?;
    current_graph(&graph, &graph_path)?;
    graph
        .reading_note_update(&note_id, &expected_revision, &body)
        .map_err(|e| e.to_string())
}

#[tauri::command(rename_all = "camelCase")]
pub fn reading_note_reattach(
    state: State<crate::AppState>,
    graph_path: String,
    note_id: String,
    expected_revision: String,
    source_page_id: String,
    selection: Option<ReadingSelection>,
) -> Result<ReadingNote, String> {
    let graph = state.graph.lock().map_err(|e| e.to_string())?;
    current_graph(&graph, &graph_path)?;
    graph
        .reading_note_reattach(
            &note_id,
            &expected_revision,
            &source_page_id,
            selection.as_ref(),
        )
        .map_err(|e| e.to_string())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn reading_notes_require_current_graph_and_camel_case_contract() {
        let directory = tempfile::tempdir_in(".").unwrap();
        let graph = Graph::open(directory.path()).unwrap();
        let other = tempfile::tempdir_in(".").unwrap();
        assert!(current_graph(&graph, other.path().to_str().unwrap()).is_err());
        assert!(current_graph(&graph, "").is_err());
        current_graph(&graph, directory.path().to_str().unwrap()).unwrap();
        let selection: ReadingSelection = serde_json::from_value(serde_json::json!({
            "pageId": "source", "blockIds": ["block"], "text": "😀",
            "kind": "source", "parts": [{
                "blockId": "block", "text": "😀", "from": 0, "to": 2, "prefix": "", "suffix": ""
            }], "documentRange": { "from": 2, "to": 4, "text": "😀" }
        }))
        .unwrap();
        assert_eq!(selection.parts[0].to, 2);
        let page = graph
            .create_page_with_content("Synthetic", false, "- Before\n")
            .unwrap();
        let note = graph
            .reading_note_create(&uuid::Uuid::new_v4().to_string(), &page.id, None, "Body")
            .unwrap();
        let json = serde_json::to_value(&note).unwrap();
        for field in [
            "notePageId",
            "filePath",
            "statusMessage",
            "targetBlockId",
            "createdAt",
            "updatedAt",
        ] {
            assert!(json.get(field).is_some(), "{field}");
        }
        assert_eq!(json["source"]["pageId"], page.id);
        assert_eq!(json["status"], "attached");
        assert_eq!(json["storage"], "inline");
        assert_eq!(json["footnoteLabel"], "grafium-note-1");
        assert_eq!(json["notePageId"], page.id);
        assert!(json["noteBlockId"].is_string());
        assert!(json["targetBlockId"].is_null());
        assert_eq!(
            serde_json::to_value(graph.reading_notes_list(None).unwrap()).unwrap()["notes"][0]
                ["id"],
            note.id
        );
    }
}
